cmd/ignite/run/exec.go

```diff
@@ -0,0 +1,154 @@
+package run
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"os"
+	"path"
+	"time"
+
+	api "github.com/weaveworks/ignite/pkg/apis/ignite"
+	"github.com/weaveworks/ignite/pkg/constants"
+	"github.com/weaveworks/ignite/pkg/util"
+
+	"golang.org/x/crypto/ssh"
+	shellescape "gopkg.in/alessio/shellescape.v1"
+)
+
+type ExecFlags struct {
+	Timeout      uint32
+	IdentityFile string
+}
+
+type execOptions struct {
+	*ExecFlags
+	vm      *api.VM
+	command []string
+}
+
+func (ef *ExecFlags) NewExecOptions(vmMatch string, command ...string) (eo *execOptions, err error) {
+	eo = &execOptions{
+		ExecFlags: ef,
+		command:   command,
+	}
+	eo.vm, err = getVMForMatch(vmMatch)
+	return
+}
+
+func Exec(eo *execOptions) error {
+	// Check if the VM is running
+	if !eo.vm.Running() {
+		return fmt.Errorf("VM %q is not running", eo.vm.GetUID())
+	}
+
+	// Get the IP address
+	ipAddrs := eo.vm.Status.IPAddresses
+	if len(ipAddrs) == 0 {
+		return fmt.Errorf("VM %q has no usable IP addresses", eo.vm.GetUID())
+	}
+
+	// If an external identity file is specified, use it instead of the internal one
+	privKeyFile := eo.IdentityFile
+	if len(privKeyFile) == 0 {
+		privKeyFile = path.Join(eo.vm.ObjectPath(), fmt.Sprintf(constants.VM_SSH_KEY_TEMPLATE, eo.vm.GetUID()))
+		if !util.FileExists(privKeyFile) {
+			return fmt.Errorf("no private key found for VM %q", eo.vm.GetUID())
+		}
+	}
+	signer, err := newSignerForKey(privKeyFile)
+	if err != nil {
+		return fmt.Errorf("unable to create signer for private key: %v", err)
+	}
+
+	// Create an SSH client, and connect, we will use this to exec
+	config := newSSHConfig(signer, eo.Timeout)
+	client, err := ssh.Dial("tcp", net.JoinHostPort(ipAddrs[0].String(), "22"), config)
+	if err != nil {
+		return fmt.Errorf("failed to dial: %v", err)
+	}
+
+	// run the command, DO NOT wrap this error as the caller can check for the command exit
+	// code in the ssh.ExitError type
+	return runSSHCommand(client, eo.command)
+}
+
+func newSignerForKey(keyPath string) (ssh.Signer, error) {
+	key, err := ioutil.ReadFile(keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read private key: %v", err)
+	}
+
+	// Create the Signer for this private key.
+	return ssh.ParsePrivateKey(key)
+}
+
+func newSSHConfig(publicKey ssh.Signer, timeout uint32) *ssh.ClientConfig {
+	return &ssh.ClientConfig{
+		User: "root",
```

Comment:
Is there a scenario where `root` may not be the user wanted for ssh - VMs with "disable root login"?

Reply:
I think so, but for the moment this matches `ignite ssh` which also assumes this.

Reply:
The disable root login option default in most distributions prevents password-based authentication, but SSH keys should work just fine. We should also open an issue about allowing user selection in `ssh` and `exec`.